crates/agent/src/controllers/mod.rs

```diff
@@ -326,6 +327,44 @@ impl NextRun {
     }
 }
 
+/// Reduces a collection of results into a single result, setting the next run
+/// time to the smallest value among all the results, regardless of whether the
+/// result is an error or not.
+pub fn coalesce_results(
+    results: impl IntoIterator<Item = anyhow::Result<Option<NextRun>>>,
+) -> anyhow::Result<Option<NextRun>> {
+    let mut min = None;
+    let mut errs = Vec::new();
+    for result in results {
+        match result {
+            Ok(nr) => {
+                min = NextRun::earliest([min, nr]);
+            }
+            Err(e) => match e.downcast::<RetryableError>() {
+                Ok(rt) => {
+                    min = NextRun::earliest([min, rt.retry]);
+                    errs.push(rt.inner);
+                }
+                Err(reg_err) => errs.push(reg_err),
+            },
+        }
+    }
+    let res = if errs.is_empty() {
+        Ok(min)
+    } else if errs.len() > 1 {
+        Err(anyhow::anyhow!(
+            "{} errors:\n- {}",
+            errs.len(),
+            errs.iter().join("\n- ")
+        ))
+    } else {
+        Err(errs.into_iter().next().unwrap())
```

Comment:
```suggestion
        Err(errs.first().unwrap())
```

Reply:
`first` is only for slices and vecs, and this is just an `IntoIterator`.

Reply:
We have `let mut errs = Vec::new()` above, so it is a vector, no?